Fix `Invoke-Command` missing error on session termination.

PaulHigin opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

This change fixes a problem in Invoke-Command against a remote session and that session is abruptly terminated, but no error is reported.

## PR Context

In certain conditions, an abrupt session termination results in an Invoke-Command pipeline state going to 'Stopped' with an exception.  Currently this is ignored in job processing because a user initiated stop is not an error.  But a stopped state due to an error that is not 'PipelineStoppedException' (such as a remote transport exception) should not be ignored, but instead be treated as an error so that Invoke-Command will report it.

The fix is to update job error processing to correctly handle this error state.

Repro steps
-------------
```powershell
$session = New-PSSession -cn vm1
Invoke-Command -Session $session -ScriptBlock { 1..1000 | % { sleep 1; "Output $_" } }

# Turn off virtual machine vm1

# Result
# No error is reported by Invoke-Command

# Expected
OpenError: [vm1] Processing data from remote server vm1 failed with the following error message: The request for the Windows Remote Shell with ShellId 8E31C82A-7B31-41F0-B882-E60EF7634D33 failed because the shell was not found on the server. Possible causes are: the specified ShellId is incorrect or the shell no longer exists on the server. Provide the correct ShellId or create a new shell and retry the operation. For more information, see the about_Remote_Troubleshooting Help topic.
```

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

PaulHigin committed:
Fix Invoke-Command missing error on session termination.

PaulHigin commented:
This is a possible breaking change.  Some job ending states that used to be 'Stopped' with Exception, will now be 'Failed' with Exception.  But I feel this is the correct behavior.  The only time a 'Stopped' state is not an error is if there is no associated exception, or the exception is 'PipelineStopped'.

PoshChan commented on src/System.Management.Automation/engine/remoting/client/Job.cs:
```diff
@@ -3382,7 +3382,10 @@ internal ErrorRecord FailureErrorRecord
                                 fullyQualifiedErrorId, ErrorCategory.OpenError,
```
@PaulHigin, your last commit had 1 failures in `PowerShell-CI-windows`
Enter-PSHostProcess tests.By CustomPipeName.Can enter, exit, and re-enter using CustomPipeName
```powershell
Expected $true, because The script was able to re-enter another process and grab the pipe of 'wrpln4rk.bo1'., but got $false.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Core\Enter-PSHostProcess.Tests.ps1: line 187
187:                     Should -BeTrue -Because "The script was able to re-enter another process and grab the pipe of '$pipeName'."
```

PaulHigin replied:
@PoshChan  Please retry Windows

PaulHigin commented:
@PoshChan  Please retry windows

PoshChan commented:
@PaulHigin, successfully started retry of `PowerShell-CI-Windows`

PaulHigin committed:
Fix CodeFactor issue

JamesWTruher reviewed (commented):
are there tests that can be added for this?

PaulHigin commented:
@JamesWTruher  No, the repro is difficult to automate and still get the right error path.  We don't need another fragile test.

SteveL-MSFT commented:
@PowerShell/powershell-committee reviewed this and agree that the scenario where the pipeline is stopped with an error that is not a PipelineStopException it is a failure case, so this change is accepted.

PaulHigin commented:
@TravisEz13  This change created a subtle regression.  Please hold off merging until I have pushed the fix.

PaulHigin committed:
Fix regression, and update Stop-Job tests to wait until job is running

PaulHigin commented:
@SteveL-MSFT , @JamesWTruher  Please re-review the changes.  I have added a fix to the regression and also fixed the Stop-Job tests so that they will catch bad final state.

The regression was due to the pipeline returning a RemoteException exception on a pipeline stop instead of a PipelineStoppedException exception I was previously checking.  A pipeline stopped RemoteException will contain an ErrorRecord with FQEID of 'PipelineStopped'.

The Stop-Job tests were not catching the incorrect final state because they ran before the job was actually running.  Fix is to wait until job is running and returning data to the client.

SteveL-MSFT commented on test/powershell/engine/Job/Jobs.Tests.ps1:
```diff
@@ -311,8 +311,17 @@ Describe 'Basic Job Tests' -Tags 'Feature' {
-            # '-Seconds 100' is chosen to be substantially large, so that the job is in running state when Stop-Job is called.
-            $jobToStop = Start-Job -ScriptBlock { Start-Sleep -Seconds 100 } -Name 'JobToStop'
+            # 20 seconds is chosen to be large, so that the job is in running state when Stop-Job is called.
+            $jobToStop = Start-Job -ScriptBlock {
```
There's 4 test cases, but BeforeAll only runs once.  Should you start the test job in BeforeEach?

PaulHigin replied:
Yes, good point.  I'll also add another condition to the running wait loop.

PaulHigin committed:
CR comment: start new job for each test

PaulHigin commented:
@SteveL-MSFT  Please re-review.

PaulHigin commented:
@SteveL-MSFT  ping...
